Integration Tests Passing & Latency/LoggedOn Properties Added

jondeweydev opened the pull request:
### Summary

- Integration tests are passing and adding project queries to the web app's backend once test server sends mock response back to the 'client'.
- Queries now have the properties latency and loggedOn
- Functionality to get timestamp at beginning of middleware chain, loggedOn at end, and latency calculated between the two added.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### Issues

closes #16 
closes #19 
closes #18 
closes #21 

### Evidence

<img width="357" alt="Screen Shot 2022-06-30 at 6 08 54 PM" src="https://user-images.githubusercontent.com/95050440/176805456-84262821-06db-4798-bb77-5efc17c764e7.png">
<img width="295" alt="Screen Shot 2022-06-30 at 6 09 40 PM" src="https://user-images.githubusercontent.com/95050440/176805464-10c8860d-3d8b-4f94-a12d-1effc1393b08.png">

jondeweydev committed:
package.json fixed

jondeweydev committed:
Merge branch 'dev' of https://github.com/oslabs-beta/Gate-Logger into dev

jondeweydev committed:
removed comments and updated imports

jondeweydev committed:
beginning jest skeleton

jondeweydev committed:
updated postQuery to include success boolean instead of blocked

jondeweydev committed:
beginnings of supertest framework

jondeweydev committed:
testing errors

jondeweydev committed:
added test server

jondeweydev committed:
res.locals.graphQLGate still undefined

jondeweydev committed:
testing progression

jondeweydev committed:
logger now logs without fail

jondeweydev committed:
small changes

jondeweydev committed:
small changes

jondeweydev committed:
added functionality for logged_on, latency, and timestamp

jondeweydev committed:
updated mock data to populate db with random data

jondeweydev committed:
fixed linting errors

shalarewicz commented on src/index.ts:
```diff
@@ -29,32 +29,45 @@ const gateURI = 'http://localhost:3000';
+    const verify = newAuth.verification;
+
+    // run the API Key verification process when gatelog is instantiated
+    validate().catch((err) => err);
```
These catches isn't really doing anything and the function will still return middleware regardless of whether the verification is successful. I think we should only return middleware if the API key verification is successful, if not we should throw an error letting the developer know the API key is invalid.

jondeweydev replied:
changed it so that both of these functions either return true or an error, and if they do not return true, we return an error instead of returning the middleware callback

shalarewicz commented on src/index.ts:
```diff
@@ -29,32 +29,45 @@ const gateURI = 'http://localhost:3000';
+            const latency: number = loggedOn - timestamp;
+
+            // passes into PostQuery class all query data along with recorded timestamp, loggedOn, and latency
+            const postQuery = new PostQuery(gateURI, projectID, {
```
What is the benefit of create a new `PostQuery` object? Are we able to just call a single `post` function as opposed to instantiated a PostQuery class and then calling post()?

jondeweydev replied:
changed to an exported function instead and tested, everything is working

shalarewicz commented on src/middleware/PostQuery.ts:
```diff
@@ -21,57 +22,69 @@ export default class PostQuery {
+        // needs depth property once added into limiter functionality
+        const { complexity, tokens, success, timestamp, loggedOn, latency } = this.queryData;
 
+        // most likely redundant, merely in place to fight possible edge cases
```
Agreed, it's redundant but can't hurt to check.

shalarewicz commented on test/middleware/APIAuth.test.ts:
```diff
@@ -12,7 +12,7 @@ const MOCK_PROJECT_ID = '62997af7a5aab6a6df935797';
 const MOCK_URI = 'http://localhost:3000';
 
-describe('Test API key header verification', () => {
+xdescribe('Test API key header verification', () => {
```
Why is this test skipped?

jondeweydev replied:
During testing of the index file, just so that testing would go by faster. Will change back for the final review

shalarewicz commented on test/index.test.ts:
```diff
@@ -0,0 +1,32 @@
+                });
+        });
+    });
+});
```
I think it's worthwhile to add tests for when bad data is passed in or a connection fails to make sure errors don't disrupt the middleware chain.

jondeweydev replied:
will take care of this

jondeweydev replied:
added

shalarewicz commented on test/middleware/PostQuery.test.ts:
```diff
@@ -1,7 +1,8 @@
+import { QueryData } from '../../@types/log';
 
-describe('Test API key header verification', () => {
+xdescribe('Test API key header verification', () => {
```
Why is this test skipped?

jondeweydev replied:
undone

shalarewicz commented on test/server.ts:
```diff
@@ -0,0 +1,39 @@
+
+/**
+ * in theory, this res.send() should be calling not only res.end, but
+ * also the functionality added to res.end by the logger middleware
```
Update this comment to reflect the change to using an event listener?

jondeweydev replied:
done

shalarewicz reviewed (commented):
Good work! Some minor comments but nothing major

jondeweydev committed:
changed PostQuery.ts to only have an exported function instead of a class

jondeweydev committed:
updated error handling with APIAuth

jondeweydev committed:
fixed lint errors and removed test skipping

jondeweydev committed:
updated comments to reflect most recent changes

jondeweydev committed:
fixed TS errors related to async

jondeweydev committed:
added error throw

jondeweydev committed:
defining tests to scope out error handling

evanmcneely commented on src/index.ts:
```diff
@@ -19,42 +19,67 @@ const gateURI = 'http://localhost:3000';
+    const verify = newAuth.verification;
+
+    // run the API Key verification process when gatelog is instantiated
+    validate()
```
This could be more DRY. Can you combine `validate` and `verify` into one function? Or use `Promise.all()` to run both functions and throw the error if one of them is `false`?

jondeweydev replied:
just saw this and took care of it

evanmcneely commented on src/middleware/APIAuth.ts:
```diff
@@ -39,35 +39,41 @@ export default class AuthVerification {
         // this endpoint returns the associated project's API key
-        await axios(`${this.gateURI}/auth/${this.projectID}`)
+        await axios(`${this.gateURI}/key/${this.projectID}`)
             // .then((data: any) => data?.json())
```
You can remove this commented out `.then`

jondeweydev committed:
added new type, tweaked tests, and added new tests for index

jondeweydev committed:
fixed lint errors

jondeweydev committed:
one last fix

jondeweydev committed:
combined validate and verify into one promise.all in index

jondeweydev committed:
finishing touches on index

evanmcneely commented on test/middleware/PostQuery.test.ts:
```diff
@@ -28,25 +30,24 @@ describe('Test API key header verification', () => {
         await expect(newPost).rejects.toThrow(`[gatelog] Query data cannot be negative.`);
     });
 
     test('Error thrown when resulting projectID of query does not match the one entered', async () => {
```
This test title reads like you are expecting an error to be thrown but then you're expecting an an error not to be thrown on line 44? What is the real intention?

jondeweydev replied:
deleted this test

evanmcneely commented on test/middleware/PostQuery.test.ts:
```diff
@@ -28,25 +30,24 @@ describe('Test API key header verification', () => {
-        const newPost = () => postQuery.post();
+        const newPost = () => postQuery(mockURI, mockProjectID, mockQueryData);
 
         await expect(newPost).rejects.toThrow(`[gatelog] Query data cannot be negative.`);
```
On my machine this test is failing because for the following reason
```
Received promise resolved instead of rejected
    Resolved to value: [SyntaxError: [gatelog] Query data cannot be negative]
```
It's resolving to the right error but not rejecting.

jondeweydev replied:
fixed

evanmcneely commented on test/middleware/PostQuery.test.ts:
```diff
@@ -28,25 +30,24 @@ describe('Test API key header verification', () => {
         );
     });
 
     test('No error thrown when query data is correct syntax', async () => {
```
This is the same test as above but your but your not expecting any errors rather than a specific one. This test covers the above one, you don't need both.

jondeweydev replied:
fixed

evanmcneely commented on test/middleware/PostQuery.test.ts:
```diff
@@ -1,11 +1,11 @@
-    let postQuery: PostQuery;
+    let mockQueryData: LogQueryData;
 
     beforeEach(() => {
```
Maybe this is why one of the tests below is failing for me. The tests should be agnostic to my local setup or move these to the `.env` file so I can set up the test environment on my machine without overriding these values for everyone.

Do I need to be running the web app on on Port 3000 to run the tests?

jondeweydev replied:
yes, these tests depend on running the webapp

jondeweydev committed:
fixed all testing suite errors

jondeweydev committed:
fixed lint errors

jondeweydev commented:
Fixed testing suite, all are passing and removed unnecessary tests. Feel free to test these out yourselves just keep in mind the web application's backend needs to be running for a majority of these to pass

evanmcneely commented:
It looks like this PR may close issue #23 #17 #7 and #6 as well?

evanmcneely reviewed (approved):
Review issues associated with this PR. You may be able to link more issues and close them than is currently set up. 

Note: you need to put "closes" before each issue number for them to close automatically.

jondeweydev commented:
closes #23 closes #16 closes #21 closes #17 closes #7 closes #6
